sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala

```diff
@@ -0,0 +1,314 @@
+        isBroadcastable = false))
+  }
+
+  // scalastyle:off
+  /**
+   * The number of rows of A inner join B on A.k1 = B.k1 is estimated by this basic formula:
+   * T(A IJ B) = T(A) * T(B) / max(V(A.k1), V(B.k1)), where V is the number of distinct values of
+   * that column. The underlying assumption for this formula is: each value of the smaller domain
+   * is included in the larger domain.
+   * Generally, inner join with multiple join keys can also be estimated based on the above
+   * formula:
+   * T(A IJ B) = T(A) * T(B) / (max(V(A.k1), V(B.k1)) * max(V(A.k2), V(B.k2)) * ... * max(V(A.kn), V(B.kn)))
+   * However, the denominator can become very large and excessively reduce the result, so we use a
+   * conservative strategy to take only the largest max(V(A.ki), V(B.ki)) as the denominator.
+   */
+  // scalastyle:on
+  def joinSelectivity(
+      joinKeyPairs: Seq[(AttributeReference, AttributeReference)],
+      leftStats: Statistics,
+      rightStats: Statistics): BigDecimal = {
+
+    var ndvDenom: BigInt = -1
+    var i = 0
+    while(i < joinKeyPairs.length && ndvDenom != 0) {
+      val (leftKey, rightKey) = joinKeyPairs(i)
+      // Do estimation if we have enough statistics
+      if (columnStatsExist((leftStats, leftKey), (rightStats, rightKey))) {
+        // Check if the two sides are disjoint
+        val leftKeyStats = leftStats.attributeStats(leftKey)
+        val rightKeyStats = rightStats.attributeStats(rightKey)
+        val lRange = Range(leftKeyStats.min, leftKeyStats.max, leftKey.dataType)
+        val rRange = Range(rightKeyStats.min, rightKeyStats.max, rightKey.dataType)
+        if (Range.isIntersected(lRange, rRange)) {
+          // Get the largest ndv among pairs of join keys
+          val maxNdv = leftKeyStats.distinctCount.max(rightKeyStats.distinctCount)
+          if (maxNdv > ndvDenom) ndvDenom = maxNdv
+        } else {
+          // Set ndvDenom to zero to indicate that this join should have no output
+          ndvDenom = 0
+        }
+      }
+      i += 1
+    }
+
+    if (ndvDenom < 0) {
+      // There isn't join keys or column stats for any of the join key pairs, we do estimation like
+      // cartesian product.
+      1
+    } else if (ndvDenom == 0) {
+      // One of the join key pairs is disjoint, thus the two sides of join is disjoint.
+      0
+    } else {
+      1 / BigDecimal(ndvDenom)
+    }
+  }
+
+  /** Update column stats for output attributes. */
+  private def updateAttrStats(
+      outputRows: BigInt,
+      attributes: Seq[Attribute],
+      oldAttrStats: AttributeMap[ColumnStat],
+      joinKeyStats: AttributeMap[ColumnStat]): AttributeMap[ColumnStat] = {
+    val outputAttrStats = new mutable.HashMap[Attribute, ColumnStat]()
+    val leftRows = leftStats.rowCount.get
+    val rightRows = rightStats.rowCount.get
+    if (outputRows == 0) {
+      // empty output
+      attributes.foreach(a => outputAttrStats.put(a, emptyColumnStat(a.dataType)))
```

Comment:
we may union this plan with the other plan, then empty column stats is useful. Now we don't support union estimation because it's difficult to estimate distinct count after union.